resources/lib/library.py

```diff
@@ -646,6 +655,63 @@ def _create ( self, item, allowOverrideLabel = True ):
 
         return( listitem )
 
+    def _append_service_parameter(self, path):
+        """
+        Check if an add-on uses services in the background
+        -- Only needed in case you use the Widgets --
+        So queues an additional parameter to the path, to let the add-on know that it was initiated by the skin service.
+        THIS IS TEMPORARY WORKAROUND! IS NOT A GOOD CODE PRACTICE, appropriate modifications must be made
+        to allow you to get a signal back from each add-ons to get the status of the service ready in order to invoke
+        an add-on (to get list data) in a safe way, without forcing the start of each add-on immediately when the
+        services are not yet ready.
+
+        In this case, each add-on frontend must implement a forced loop code that is activated only when it receives
+        this parameter, in order to wait for his service to be ready before completing the required operation
+        from the skin service.
+        """
+        service_parameter = 'skin_service=wait'
+        paths = [part for part in path.split('/') if part]
+        if not paths:
+            return path
+        if paths[0] != 'plugin:':
+            return path
+        if paths[1] not in self.cache_addons_has_service:
+            has_service = False
+            try:
+                addon = xbmcaddon.Addon(id=paths[1])
+                # Verify that in addon.xml, there is the service extension
+                file_path = os.path.join(addon.getAddonInfo('path'), 'addon.xml')
+                file_content = self.load_file(file_path)
+                from re import findall
+                has_service = bool(findall(r'<extension[^<>]*point="xbmc\.service"[^<>]*\/>', file_content))
```

Comment:
Script has plenty of xml parsing, including for skin addon.xml - can current code be reused? Is a findall better (f.e. quicker, easier to read)?

Reply:
in this module is not used this library so I do not understand your point. Placing the import within the function means to load the library only when used, so as not to weigh down the module,
if in the future you need this library elsewhere you can move the import to the top

Reply:
The point was asking if you had looked at the code to see if existing code could be reused. Don't worry, you've answered that spectacularly well.